runtime/io.h

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Defines API between compiled code and I/O runtime library.
+
+#ifndef FORTRAN_RUNTIME_IO_H_
+#define FORTRAN_RUNTIME_IO_H_
+
+#include "entry-names.h"
+#include <cinttypes>
+#include <cstddef>
+
+namespace Fortran::runtime {
+class Descriptor;
+class NamelistGroup;
+};
+
+namespace Fortran::runtime::IO {
+
+class IOStatementState;
+using Cookie = IOStatementState *;
+using ExternalUnit = int;
+using AsynchronousId = int;
+static constexpr ExternalUnit DefaultUnit{-1};  // READ(*), WRITE(*), PRINT
+
+extern "C" {
+
+#define IONAME(name) RTNAME(IO_##name)
+
+// These functions initiate data transfer statements (READ, WRITE, PRINT).
+// Example: PRINT *, 666 is implemented as the series of calls:
+//   Cookie cookie{BeginExternalListOutput(DefaultUnit)};
+//   OutputInteger64(cookie, 666);
+//   EndIOStatement(cookie);
+
+// Internal I/O initiation
+// Internal I/O can loan the runtime library an optional block of memory
+// in which the library can maintain state across the calls that implement
+// the internal transfer; use of these blocks can reduce the need for dynamic
+// memory allocation &/or thread-local storage.  The block must be sufficiently
+// aligned to hold a pointer.
+constexpr std::size_t RecommendedInternalIOScratchAreaBytes(
+    int maxFormatParenthesesNestingDepth) {
```

Comment:
I assume the caller doesn't always know the max paren nest depth w/o parsing the format string.  When the caller doesn't know the value at compile time, is the recommendation not to call this routine or what?

Reply:
This is a `constexpr` function meant to be used at compilation time only.  For `FORMAT()` statements and literal format character strings, we'll know the max depth because we will have validated those strings.   For variable character formats, pick a reasonable max depth.  It's okay if it's wrong, the runtime will know and will revert to using other storage.

Reply:
Format verification code does not currently retain maximum paren nesting depth, but it would be easy to do that.